simulator/controller/controller.go

```diff
@@ -0,0 +1,194 @@
+package controller
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"golang.org/x/xerrors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/metadata"
+	"k8s.io/client-go/metadata/metadatainformer"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/controller-manager/pkg/clientbuilder"
+	"k8s.io/controller-manager/pkg/informerfactory"
+	"k8s.io/klog/v2"
+	controllermanageroptions "k8s.io/kubernetes/cmd/kube-controller-manager/app/options"
+	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+)
+
+const (
+	// controllerStartJitter is the Jitter used when starting controller.
+	controllerStartJitter = 1.0
+)
+
+type initFunc func(controllerCtx controllerContext) error
+
+// controllerInitializersFunc is used to create a collection of initializers
+// given the loopMode.
+
+type controllerInitializersFunc func() (initializers map[string]initFunc)
+
+var _ controllerInitializersFunc = newControllerInitializers
+
+type Manager interface {
+	RunController()
+	ShutdownController()
+	ResetController()
+}
+
+type manager struct {
+	shutdownFunc func()
+	client       clientset.Interface
+	config       *restclient.Config
+}
+
+func New(client clientset.Interface, c *restclient.Config) Manager {
+	return &manager{
+		client: client,
+		config: c,
+	}
+}
+
+func (s *manager) restartController() {
+	s.shutdownFunc()
+	s.RunController()
+}
+
+func (s *manager) ResetController() {
+	s.restartController()
+}
+
+func (s *manager) ShutdownController() {
+	if s.shutdownFunc != nil {
+		s.shutdownFunc()
+	}
+}
+
+func (s *manager) RunController() {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go run(s.client, s.config, ctx.Done())
+	shutdownFunc := func() {
+		cancel()
+	}
+	s.shutdownFunc = shutdownFunc
+}
+
+// run runs the KubeControllerManagerOptions.  This should never exit.
+func run(client clientset.Interface, config *restclient.Config, stopCh <-chan struct{}) {
+	run := func(ctx context.Context, initializersFunc controllerInitializersFunc) {
+		controllerContext, err := createControllerContext(client, config, ctx.Done())
+		if err != nil {
+			klog.Fatalf("error building controller context: %v", err)
```

Comment:
We shouldn't use klog.Fatalf in application codes because it calls os.Exit internally and the defer functions will not be called.
> deferred functions are not run.
https://golang.org/pkg/os/#Exit

https://play.golang.org/p/f7IjXeb62dd

---

So, could you return errors as much as you can when something goes wrong? 
I guess only `startControllers` needs to be called within goroutine. So, in other parts, you can return errors as we normally do.